packages/filesystem/src/browser/file-resource.ts

```diff
@@ -58,6 +58,10 @@ export class FileResource implements Resource {
     }
 
     async readContents(options?: { encoding?: string }): Promise<string> {
+        const fileStat = await this.getFileStat();
+        if (!fileStat) {
+            throw new Error(`Resource for URI ${this.uriString} does not exist`);
```

Comment:
`this.fileSystem.resolveContent` already throws an exception if a file does not exist. What is the point of doing an additional request to get the same results?

Reply:
The goal is to avoid the backend outputting the error. Even with the `try/catch` in the frontend, the backend will still display the error, even when caught.

See: https://github.com/theia-ide/theia/blob/master/packages/filesystem/src/node/node-filesystem.ts#L72

The backend will emit an error, but it is actually called via proxy in the backend, so it doesn't make sense to catch it at this point. So we assumed for the error to not appear we would not call this function and test if it will throw before hand.

Reply:
it will cause 2 additional requests to a server for each read. Exception on the backend is annoying indeed, but it does not cause any harm to compromise performance.

Reply:
So should we close the issue and leave the error as it is ?

Reply:
I don't know. It is a generic issue: whenever an error is thrown from the backend, it is logged.

Is there a way to add a guard specifically for preferences code? `file-resource` is used in a lot of places.

Reply:
IMO resources should have a new method in their interface, such as isAvailable(): Promise<boolean>, because these objects are abstractions of something that could not exist (in the case of files).

Just taking a look at the implementers for the Resource interface, it is like 7 implementations. Should we default to isAvailable = () => true when we don't know what to do of it ? Right now when the resource is not available, we keep doing operations on it as if it was ok, so defaulting to true for now should be fine too ? (when changing the API)

Reply:
> IMO resources should have a new method in their interface, such as isAvailable(): Promise<boolean>

it is redundant, an exception on `read` already indicating that a resource does not exist. 

There is nothing wrong with the resource API. As I mentioned above the issue is general: whenever `filesystem.resolveContent` is called if a file does not exist an error will be logged on the backend. Regardless whether is called from `FileResource` or other code.

I see several alternatives:
- (1) don't do anything, we don't want to expose sever stack traces to clients, logging it is ok. Users never see these errors, only developers and devops, plus errors are not harmful.
- (2) add a guard only for the case when `stat` is `undefined`, a guard should only call `exists` to minimize the number of requests. If `stat` is defined then leave the code as it is. It is a minimal change to fix #2211 without impacting perf.
- (3) change `resolveContent` that it does not throw but returns an empty object if a file does not exist. It is very impactful change. Considering that #2211 is a questionable issue from the user perspective, i would rather do (2) or (1).

Reply:
```ts
// file-resource.ts

    async readContents(options?: { encoding?: string }): Promise<string> {
        const { stat, content } = await this.fileSystem.resolveContent(this.uriString, options);
        this.stat = stat;
        return content;
    }
```
In this method, trying to test `stat` would be too late, because the backend would have thrown already.

What about add a `silent?: boolean` option to the `resolveContent` protocol API ?

We would only set `silent: true` when we actually catch the exception from the frontend. And it would only catch our custom generated errors:
```ts
    async resolveContent(uri: string, options?: { encoding?: string, silent?: boolean }): Promise<{ stat: FileStat, content: string }> {
        const _uri = new URI(uri);
        const stat = await this.doGetStat(_uri, 0);
        if (!silent) {
            if (!stat) {
                throw new Error(`Cannot find file under the given URI. URI: ${uri}.`);
            }
            if (stat.isDirectory) {
                throw new Error(`Cannot resolve the content of a directory. URI: ${uri}.`);
            }
        }
        const encoding = await this.doGetEncoding(options);
        const content = await fs.readFile(FileUri.fsPath(_uri), { encoding });
        return { stat, content };
    }
```
The only impact this would have on performance would be the extra option `silent` added in the jsonrpc message when we know that we catch the error in the frontend.

Reply:
Ok, just realized the `silent` proposal was silly, because we have to throw something to stop the function and notify the frontend, but it will get logged.

But in order to clean the console a bit, how about just logging something like:
```ts
        if (!stat) {
            throw { message: `Cannot find file under the given URI. URI: ${uri}.`};
        }
        if (stat.isDirectory) {
            throw { message: `Cannot resolve the content of a directory. URI: ${uri}.`};
        }
```
This will just display:
```
root ERROR Request resolveContent failed with error: Cannot find file under the given URI. URI: file:///home/<me>/<project>/.theia/settings.json.
```
Because do we want to keep the stacktrace anyway ?

Reply:
Unless you have an idea, I guess we would have to let things as they are now, with the full error being logged.